### PR TITLE
Add `FullTransformationMonoid` and `PartialTransformationMonoid` presentations

### PR DESCRIPTION
**Pull request by MTWhyte**
This PR adds a constructions for presentations of the full and partial transformation monoids.

### CODE REVIEW

**Comment by MTWhyte** [comment]
If I run `REQUIRE(p.rules != p.rules)` before creating the ToddCoxeter on line `2628` of `tests/test-todd-coxeter.cpp`, I retrieve the following (modulo formatting) for `p.rules` in the error output:

```
{ { 0, 0 }, { 5 },
  { 1, 1 }, { 5 },
  { 2, 2 }, { 5 },
  { 0, 1, 0, 1, 0, 1 }, { 5 },
  { 1, 2, 1, 2, 1, 2 }, { 5 },
  { 2, 0, 2, 0, 2, 0 }, { 5 },
  { 0, 1, 0, 2, 0, 1, 0, 2 }, { 5 },
  { 1, 2, 1, 0, 1, 2, 1, 0 }, { 5 },
  { 2, 0, 2, 1, 2, 0, 2, 1 }, { 5 },
  { 3, 3 }, { 3 },
  { 3, 0, 3, 0 }, { 0, 3, 0, 3 },
  { 0, 3, 0, 1 }, { 1, 0, 3, 0 },
  { 1, 3, 1, 0 }, { 0, 1, 3, 1 },
  { 0, 3, 0, 2 }, { 2, 0, 3, 0 },
  { 2, 3, 2, 0 }, { 0, 2, 3, 2 },
  { 0, 3, 0, 3, 0 }, { 0, 3, 0, 3 },
  { 4 }, { 4, 4 },
  { 4, 0, 3, 0 }, { 4 },
  { 0, 3, 0, 4 }, { 0, 3, 0 },
  { 4, 3 }, { 0, 3, 0, 3, 4 },
  { 4, 1, 3, 1 }, { 1, 3, 1, 4 },
  { 0, 5 }, { 0 },
  { 5, 0 }, { 0 },
  { 1, 5 }, { 1 },
  { 5, 1 }, { 1 },
  { 2, 5 }, { 2 },
  { 5, 2 }, { 2 },
  { 3, 5 }, { 3 },
  { 5, 3 }, { 3 },
  { 4, 5 }, { 4 },
  { 5, 4 }, { 4 },
  { 5, 5 }, { 5 } }
```

As relates to Theorem 9.4 of *Classical Finite Transformation Semigroups*, the generators `0`, `1`, `2`, `3`, `4` and `5` correspond to pi_2, pi_3, pi_4, epsilon_(1), epsilon_{1, 2} and the identity, respectively. I have thoroughly checked the output, and I'm very confident that it is what it should be.

The relation `{{4, 4}, {4}}` (that the non-permutation full-transformation generator is idempotent) is omitted from the theorem, but I think it's clear that it should be there (and even if I was mistaken, adding that relation couldn't cause the problem).

**Comment by MTWhyte** [comment]
To be a bit more specific: the number of active cosets after about a minute is hovering at around 6 million, which is why I think there is a problem.

File: tests/fpsemi-examples.cpp [hunk 1]
```diff
@@ -1052,17 +1052,17 @@ namespace libsemigroups {
               {{3, 1, 1, 2, 1}, {3, 1, 0, 2}},
               {{1, 2, 1, 2, 0, 2}, {2, 1, 2, 0, 2}}};
     } else if (n >= 4 && val == author::Aizenstat) {
-      LIBSEMIGROUPS_EXCEPTION("not yet implemented");
-      // FIXME this is missing the relations from T_n
```
Comment (james-d-mitchell) [inline tests/fpsemi-examples.cpp:1056]: Looking at this quickly just now, have you maybe made the same mistake as me, forgotten to add the relations from the full transformation monoid mentioned in this FIXME?

Comment (MTWhyte) [reply tests/fpsemi-examples.cpp:1056]: Ah I have misread this, in a way I can't even see how I managed to -- I will fix now, thanks!

**Comment by MTWhyte** [comment]
I've now got KnuthBendix computing the correct size of the full transformation monoid of degree 4 in about 200ms. Degree 5 takes roughly a minute, but also gives the correct size.

**Comment by MTWhyte** [comment]
When building the tests, I am warned:

```
tests/fpsemi-examples.cpp:1158:3: warning: non-void function does not return a value in all control paths [-Wreturn-type]
```

I think this means it's possible for a function (either the whole method, or one of the lambdas) whose return type is not `void` to return nothing, but I've tried to fix this where I thought it could be possible. I'll keep looking, but note it here!

**Comment by MTWhyte** [comment]
ToddCoxeter is much faster. Compared to KnuthBendix's minute, it computes the correct size of the full transformation monoid of degree 5 in 300ms. For degree 6, the time for the correct size is about 16 seconds.

**Comment by MTWhyte** [comment]
Looks like the linter (at least) isn't happy, though I did run `make format` on this, and double checked just now by running it again and seeing that it didn't change any of the relevant files. I'll see if I can figure this out tomorrow.

File: tests/fpsemi-examples.cpp [hunk 2]
```diff
@@ -1052,20 +1052,108 @@ namespace libsemigroups {
...
+      auto                   result = SymmetricInverseMonoid(n, author::Sutov);
+      word_type              e12    = {n};
+      std::vector<word_type> pi;
+      for (size_t i = 0; i <= n - 2; ++i) {
+        pi.push_back({i});
+      }
+
+      // The following expresses the epsilon idempotents in terms of the
+      // generating set
+      auto eps = [=](size_t i, size_t j) -> word_type {
```
Comment (MTWhyte) [inline tests/fpsemi-examples.cpp:1066]:
```suggestion
      auto eps = [&e12, &pi, etc...](size_t i, size_t j) -> word_type {
```

Comment (MTWhyte) [reply tests/fpsemi-examples.cpp]: Done

**Comment by MTWhyte** [comment]
I've added a separate construction for the full transformation monoid, and combined the then-two functions for full transformation monoid presentations into one. The author value for the one I've added is `author::Iwahori`.

File: tests/fpsemi-examples.cpp [hunk 3]
```diff
@@ -99,6 +99,115 @@ namespace libsemigroups {
...
+        } else if (i == 2) {
+          return pi[j - 2] * pi[0] * e12 * pi[0] * pi[j - 2];
+        }
+        return pi[i - 2] * pi[0] * pi[j - 2] * pi[0] * e12 * pi[0] * pi[j - 2]
+               * pi[0] * pi[i - 2];
+      };
+
+      auto transp = [&pi](size_t i, size_t j) -> word_type {
+        LIBSEMIGROUPS_ASSERT(i != j);
+        if (i > j) {  // FIXME Want to do: return transp(j, i);, but can't do
```
Comment (MTWhyte) [inline tests/fpsemi-examples.cpp:151]: *Todo:* I'd prefer to achieve the same here with `return transp(j, i)`, but this won't compile. I understand this is something to do with the lambda function not being within its own scope, and I'll figure out what the best way to sort this is.

Comment (MTWhyte) [reply tests/fpsemi-examples.cpp]: I understand that I can fix this by adding `transp` as a third argument to itself, per [this](https://artificial-mind.net/blog/2020/09/12/recursive-lambdas) solution. However, I think this is way less readable than the temporary 'manual' swap I put in, so I won't. I could also get away with not setting `i < j`, but I think doing this makes the relations easier to read.

I'll leave it as is for now.

File: tests/fpsemi-examples.cpp [hunk 4]
```diff
@@ -1106,7 +1226,8 @@ namespace libsemigroups {
   }
 
   // Chinese monoid
-  // See: The Chinese Monoid - Cassaigne, Espie, Krob, Novelli and Hivert, 2001
+  // See: The Chinese Monoid - Cassaigne, Espie, Krob, Novelli and Hivert,
+  // 2001
```
Comment (MTWhyte) [inline tests/fpsemi-examples.cpp:1230]: This is an accidental change from where the formatter alters a file I've modified, but in a place where I've not modified it. I can remove this if that's best.

File: tests/fpsemi-examples.cpp [hunk 5]
```diff
@@ -99,6 +99,114 @@ namespace libsemigroups {
...
+        return pi[i - 2] * pi[0] * pi[j - 2] * pi[0] * e12 * pi[0] * pi[j - 2]
+               * pi[0] * pi[i - 2];
+      };
+
+      auto transp = [&pi](size_t i, size_t j) -> word_type {
+        LIBSEMIGROUPS_ASSERT(i != j);
+        if (i > j) {
+          size_t k = i;
+          i        = j;
+          j        = k;
```
Comment (james-d-mitchell) [inline tests/fpsemi-examples.cpp:154]:
```suggestion
   std::swap(i, j);
```

Comment (james-d-mitchell) [reply tests/fpsemi-examples.cpp:154]: With the correct amount of indentation!

File: tests/fpsemi-examples.cpp [hunk 6]
```diff
@@ -99,6 +99,114 @@ namespace libsemigroups {
...
+      std::vector<word_type> pi;
+      for (size_t i = 0; i <= n - 2; ++i) {
+        pi.push_back({i});
+      }
+
+      // The following expresses the epsilon idempotents in terms of the
+      // generating set
+      auto eps = [&e12, &pi](size_t i, size_t j) -> word_type {
+        LIBSEMIGROUPS_ASSERT(i != j);
+        if (i == 1 and j == 2) {
```
Comment (james-d-mitchell) [inline tests/fpsemi-examples.cpp:132]:
```suggestion
        if (i == 1 && j == 2) {
```

Comment (james-d-mitchell) [reply tests/fpsemi-examples.cpp:132]: Although the keywords `and`, `or`, and `not` exist in C++, we use `&&`, `||` and `!` instead.

File: tests/fpsemi-examples.cpp [hunk 7]
```diff
@@ -99,6 +99,114 @@ namespace libsemigroups {
...
+      // This function adds the full transformation monoid relations due to
+      // Iwahori, from Section 9.3, p161-162, (Ganyushkin + Mazorchuk),
+      // expressed in terms of the generating set {pi_2, ..., pi_n,
+      // epsilon_{12}} using the notation of that chapter.
+      // https://link.springer.com/book/10.1007/978-1-84800-281-4
+
+      // The argument m specifies the letter value the idempotent e12
+      // corresponds to. When adding these relations for the full transformation
+      // monoid presentation, we want m = n - 1. For the partial transformation
+      // monoid presentaiton, we want m = n.
```
Comment (james-d-mitchell) [inline tests/fpsemi-examples.cpp:115]:
```suggestion
      // monoid presentation, we want m = n.
```

File: tests/fpsemi-examples.cpp [hunk 8]
```diff
@@ -99,6 +99,114 @@ namespace libsemigroups {
         }
       }
     }
+    void add_Iwahori_full_transformation_monoid_relations(
```
Comment (james-d-mitchell) [inline tests/fpsemi-examples.cpp:102]:
```suggestion
    void add_iwahori_full_transformation_monoid_relations(
```
We use snake case for functions (mostly) which doesn't include caps.

File: tests/fpsemi-examples.cpp (hunk 4 above)
Comment (james-d-mitchell) [reply tests/fpsemi-examples.cpp:1230]: Yes please.

**Review by james-d-mitchell — APPROVED** [review]
Couple of minor things to fix, then I'm happy to merge this!

Thanks! @MTWhyte

File: tests/fpsemi-examples.cpp (hunk 4 above)
Comment (MTWhyte) [reply tests/fpsemi-examples.cpp:1230]: Done

File: tests/fpsemi-examples.cpp (hunk 6 above)
Comment (MTWhyte) [reply tests/fpsemi-examples.cpp:132]: Done

File: tests/fpsemi-examples.cpp (hunk 7 above)
Comment (MTWhyte) [reply tests/fpsemi-examples.cpp:115]: Done

File: tests/fpsemi-examples.cpp (hunk 8 above)
Comment (MTWhyte) [reply tests/fpsemi-examples.cpp:102]: Done

File: tests/fpsemi-examples.cpp (hunk 5 above)
Comment (MTWhyte) [reply tests/fpsemi-examples.cpp:154]: Done

**Pull request merged**